docs/design/services/s3/transfermanager/prototype.java

```diff
@@ -0,0 +1,753 @@
+
+    /**
+     * The part number.  */
+    int partNumber();
+
+    /**
+     * The offset from the beginning of the object where this part begins.
+     */
+    long partOffset();
+
+    /**
+     * The size of the part requested in bytes.
+     */
+    long size();
+
+    /**
+     * Whether this is the last part of the object.
+     */
+    boolean isLastPart();
+}
+
+/**
+ * Context object for a single part download of an object.
+ */
+public interface SinglePartDownloadContext {
+    /**
+     * The original download request given to the Transfer Manager.
+     */
+    DownloadObjectRequest downloadRequest();
+
+    /**
+     * The request sent to S3 for this object. This is empty if downloading a presigned URL.
+     */
+    Optional<GetObjectRequest> objectRequest();
+}
+
+/**
+ * Progress listener for a Transfer.
+ * <p>
+ * The SDK guarantees that calls to {@link #transferProgressEvent(EventContext)}
+ * are externally synchronized.
+ */
+public interface TransferProgressListener {
+    /**
+     * Called when a new progress event is available for a Transfer.
+     *
+     * @param ctx The context object for the given transfer event.
+     */
+    void transferProgressEvent(EventContext ctx);
+
+    interface EventContext {
+    }
+
+    interface Initiated extends EventContext {
+        /**
+         * The transfer this listener associated with.
+         */
+        Transfer<?> transfer();
+    }
+
+    interface BytesTransferred extends Initiated {
+        /**
+         * Average number of bytes transferred per second for this transfer.
+         */
+        long averageTransferBytesPerSecond();
+
+        /**
+         * The transfer request for the object whose bytes were transferred.
+         */
+        TransferObjectRequest objectRequest();
+
+        /**
+         * The number of bytes transferred for this event.
+         */
+        long bytes();
+
+        /**
+         * The total length of the object.
+         */
+        long objectLength();
+
+        /**
+         * If the transfer of the given object is complete.
+         */
+        boolean complete();
+    }
+
+    interface Completed extends Initiated {
+    }
+
+    interface Cancelled extends Initiated {
+    }
+
+    interface Failed extends Initiated {
+        /**
+         * The error.
+         */
+        Throwable error();
+    }
+}
+
+/**
+ * Hints to help optimize an object download.
+ */
+public interface DownloadHints {
+    /**
+     * The size of the object to be downloaded. In parallel download situations
+     * this allows the Transfer Manager to avoid making an extra request to S3
+     * to find the size of the object in order to determine proper object
+     * partitioning.
+     */
+    long objectLength();
+}
+
+/**
+ * A download transfer of a single object from S3.
+ */
+public interface DownloadObject extends Transfer<DownloadObject> {
+    @Override
+    CompletableFuture<DownloadObject> completionFuture();
```

Comment:
Could we do `CompletableFuture<Void>` or something like `CompletableFuture<CompletedDownloadObject>`?

Reply:
Hmm I like the `CompletableFuture<CompletedDownloadObject>` idea :)